authz/rbac_translator.go

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package authz exposes methods to manage authorization within gRPC.
+package authz
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+)
+
+type header struct {
+	Key    *string   `json:"key"`
```

Comment:
So for encoding/json to Unmarshal the policy into struct, the fields have to be exported, so that they are visible to the json package. If we don't export, Unmarshal result would be empty.

Reply:
I guess my question is actually like, why don't we export the struct itself(e.g. `header` to be `Header`), if the whole thing is going to be used by another package?
The only case I can think of is when sometimes we want the callers to not be able to construct the struct themselves. For example, we want the callers to only be able to 
```
h := CreateHeaders(...)
fmt.Println(h.Key) // use public fields 
```
but not 
```
h := &Headers{...}
```
Is that our case here?

Reply:
We won't need the struct outside of this source file. It is used only for Unmarshalling the Json input.